pkg/lib/packagekit.es6

```diff
@@ -302,3 +309,172 @@ export function watchRedHatSubscription(callback) {
             }
             );
 }
+
+/* Support for installing missing packages.
+ *
+ * First call check_missing_packages to determine whether something
+ * needs to be installed, then call install_missing_packages to
+ * actually install them.
+ *
+ * check_missing_packages resolves to an object that can be passed to
+ * install_missing_packages.  It contains these fields:
+ *
+ * - missing_names:     Packages that were requested, are currently not installed,
+ *                      and can be installed.
+ *
+ * - unavailable_names: Packages that were requested, are currently not installed,
+ *                      but can't be found in any repository.
+ *
+ * If unavailable_names is empty, a simulated installation of the missing packages
+ * is done and the result also contains these fields:
+ *
+ * - extra_names:       Packages that need to be installed as dependencies of
+ *                      missing_names.
+ *
+ * - remove_names:      Packages that need to be removed.
+ *
+ * - download_size:     Bytes that need to be downloaded.
+ *
+ * Both check_missing_packages and install_missing_packages have a
+ * "progress" callback parameter, like cancellableTransaction.
+ *
+ * In addition to the usual "waiting", "percentage", and "cancel" fields,
+ * the progress object reported by install_missing_packages also includes
+ * "info" and "package" from the "Package" signal.
+ */
+
+export function check_missing_packages(names, progress_cb) {
+    var data = {
+        missing_names: [ ],
+        unavailable_names: [ ],
+    };
+
+    if (names.length === 0)
+        return Promise.resolve(data);
+
+    function refresh() {
+        return cancellableTransaction("RefreshCache", [ false ], progress_cb);
+    }
+
+    function resolve() {
+        data.missing_ids = [ ];
+
+        var installed_names = { };
+
+        return cancellableTransaction("Resolve",
+                                      [ Enum.FILTER_ARCH |
+                                        Enum.FILTER_NOT_SOURCE |
+                                        Enum.FILTER_NEWEST,
+                                      names
+                                      ],
+                                      progress_cb,
+                                      {
+                                          Package: (info, package_id) => {
+                                              var parts = package_id.split(";");
+                                              var repos = parts[3].split(":");
```

Comment:
`var` → `let` in arrow functions or nested scopes please; scoping of `var` doesn't go well with either. Same problem further down.

Reply:
Fixed.  What's wrong with `var`in arrow functions?  It's as broken as with normal functions, but not more broken, right?

Reply:
No, `let` has a scope of the enclosing `{ }` block, i. e. what you intuitively expect. `var`'s scope is always the innermost enclosing `function` (*not* arrow-closure!), so that if you declare a `var` in an arrow function, `for` loop, `if` block etc., it will still exist *outside* of that block. I. e. `var` other than directly in a `function` is conceptually broken and error prone, and should be avoided. 

See https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Statements/let